### PR TITLE
Insert in-place rotation waypoint if missing

### PR DESCRIPTION
**Pull request by xiyuoh**
This PR targets https://github.com/open-rmf/rmf/issues/556 where in-place rotation commands for certain paths are missing. After tracing the fleet adapter and `rmf_traffic` planner, it seems the planner sometimes fails to provide these waypoints. These in-place rotation waypoints may be essential for AGVs that rely entirely on RMF to align towards their next destination. The fix implemented is to check for such gaps inside `reconstruct_waypoints` and insert the rotation command as needed.

**Test the fix**

The issue can be consistently reproduced in the Office demos world by sending a patrol command for `tinyRobot1` to `hardware_2`. Do set the `skip_rotation_commands` param to `False` when testing.

When the robot is returning from `hardware_2` to its charger, it would provide the following path:

```
[tinyRobot_fleet_adapter]: Executing final go_to_place [tinyRobot1_charger] for robot [tinyRobot/tinyRobot1]
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 20.89365413 -10.3083739   -1.59019108] on map [L1]: cmd_id 14
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 20.89365413 -10.3083739   -3.11109903] on map [L1]: cmd_id 15
[tinyRobot_fleet_adapter]: Opening door [hardware_door] for [tinyRobot/tinyRobot1]
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 18.79442303 -10.3724069   -3.11109903] on map [L1]: cmd_id 16
[tinyRobot_fleet_adapter]: Releasing door [hardware_door] for [tinyRobot/tinyRobot1]
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [18.73952429 -6.87398187 -1.5551052 ] on map [L1]: cmd_id 17
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 1.87395243e+01 -6.87398187e+00  3.91881687e-03] on map [L1]: cmd_id 18
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 1.68558247e+01 -6.88136378e+00  3.91881687e-03] on map [L1]: cmd_id 19
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [11.5658893  -6.99680773  0.0182071 ] on map [L1]: cmd_id 20
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [11.5658893  -6.99680773 -0.01173879] on map [L1]: cmd_id 21
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [10.08614619 -6.97943654 -0.01173879] on map [L1]: cmd_id 22
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [10.08614619 -6.97943654  1.32861949] on map [L1]: cmd_id 23
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [10.4330537  -5.57509559  1.32861949] on map [L1]: cmd_id 24
[tinyRobot_fleet_adapter]: Charging [tinyRobot/tinyRobot1] to [100.000000]
```
Notice the consecutive commands provided to the fleet adapter with `cmd_id` 19 and 20 have neither the same location nor yaw. path provided to the fleet adapter goes from 

With this PR, an intermediate rotation waypoint (`cmd_id` 20) will be inserted and provided to the fleet adapter:
```
[tinyRobot_fleet_adapter]: Executing final go_to_place [tinyRobot1_charger] for robot [tinyRobot/tinyRobot1]
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 20.89365413 -10.3083739   -1.59019246] on map [L1]: cmd_id 14
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 20.89365413 -10.3083739   -3.11109903] on map [L1]: cmd_id 15
[tinyRobot_fleet_adapter]: Opening door [hardware_door] for [tinyRobot/tinyRobot1]
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 18.79442303 -10.3724069   -3.11109903] on map [L1]: cmd_id 16
[tinyRobot_fleet_adapter]: Releasing door [hardware_door] for [tinyRobot/tinyRobot1]
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [18.73952429 -6.87398187 -1.5551052 ] on map [L1]: cmd_id 17
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 1.87395243e+01 -6.87398187e+00  3.91881687e-03] on map [L1]: cmd_id 18
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [ 1.68558247e+01 -6.88136378e+00  3.91881687e-03] on map [L1]: cmd_id 19
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [16.85582469 -6.88136378  0.0182071 ] on map [L1]: cmd_id 20
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [11.5658893  -6.99680773  0.0182071 ] on map [L1]: cmd_id 21
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [11.5658893  -6.99680773 -0.01173879] on map [L1]: cmd_id 22
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [10.08614619 -6.97943654 -0.01173879] on map [L1]: cmd_id 23
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [10.08614619 -6.97943654  1.32861949] on map [L1]: cmd_id 24
[tinyRobot_command_handle]: Commanding [tinyRobot1] to navigate to [10.4330537  -5.57509559  1.32861949] on map [L1]: cmd_id 25
[tinyRobot_fleet_adapter]: Charging [tinyRobot/tinyRobot1] to [100.000000]
```

### CODE REVIEW

**Comment by osama-z-salah** [comment]
Hi @xiyuoh , as mentioned in https://github.com/open-rmf/rmf/issues/556, this problem consistently occurs in the following scenarios:

1. After the fleet adapter plans a new go_to_place location.
2. After the robot finishes waiting for traffic.

Your PR has resolved the issue for the first scenario, but the problem still persists after the robot done waiting for traffic where in-place rotation commands for certain paths are missing.

**Comment by xiyuoh** [comment]
Hi @osama-z-salah , thanks for testing this PR! I've tried out what you're describing in sim and it doesn't seem like its related to traffic, but I do see in-place rotations being skipped from within the EasyFullControl fleet adapter. I've opened another [PR](https://github.com/open-rmf/rmf_ros2/pull/398) for that since the code lives in another repo, feel free to try that out and let me know if the issue persists.

File: rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp
Comment (xiyuoh) [reply rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp:562]: I think same graph index always mean same position, but same (or sufficiently close) positions don't always mean same graph index. Having  `!(same_stack || same_pos)` was to include any locations that either have the same graph index, or a different graph index but sufficiently close within the threshold of 0.001m. I can see cases where two separate waypoints are added close to one another to manage the lane direction and traffic.

File: rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp
Comment (xiyuoh) [reply rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp]: [53908a2](https://github.com/open-rmf/rmf_traffic/pull/111/commits/53908a2650917c52015e2088fab37381bb1856fe)

File: rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp
Comment (luca-della-vedova) [reply rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp]: Yap that's what I meant, if two waypoints have the same index they will always have the exact same position, so would checking only for the position be sufficient?

Comment (xiyuoh) [reply rmf_traffic/src/rmf_traffic/agv/planning/DifferentialDrivePlanner.cpp]: Ah I see what you mean, gotcha removed in [67ae677](https://github.com/open-rmf/rmf_traffic/pull/111/commits/67ae677c73ebe88edb373b9221a32a513fea607c)

**Review by luca-della-vedova — APPROVED** [review]
Thanks for iterating! Someone with admin access needs to go through the required workflows for this repo though, or this will be blocked on CI settings

**Review by mxgrey — APPROVED** [review]
After fixing the CI with https://github.com/open-rmf/rmf_traffic/pull/112 it turns out that there were some test failures in the PR. There were two root causes:

1. The "same orientation" calculation wasn't taking into account overflow at 2*pi
2. We weren't matching the plan waypoint with a checkpoint in the trajectory

Both issues are fixed by [344559c](https://github.com/open-rmf/rmf_traffic/pull/111/commits/344559c3e20a992afa8a3357796b6084a8fa0af1). (2) probably doesn't matter much to be honest, but I thought of a semi-reasonable way to make it work, and that spares us from needing to put fudge into the assumptions of the unit tests.

I'd appreciate if someone can take some time to test these changes on a demo world before merging. I don't think (2) should have any detrimental effects, but it's better to do smoke test on that than to assume.

**Comment by xiyuoh** [comment]
Thanks for fixing the remaining issues! Ran this PR a couple times in office world and things are looking good, will merge it in.

**Pull request merged**
